web/src/lib/actions/input.ts

```diff
@@ -0,0 +1,229 @@
+export type ShortcutCallback = (event: KeyboardEvent) => false | unknown;
+
+export type KeyTargets = HTMLElement | Document | Window;
+export type KeyDownListenerFactory = (element: KeyTargets) => (event: KeyboardEvent) => void;
+
+export type KeyCombo = {
+  key: string;
+  alt?: boolean;
+  ctrl?: boolean;
+  shift?: boolean;
+  meta?: boolean;
+};
+export type KeyInput = string | string[] | KeyCombo | KeyCombo[];
+
+// OS/Browser well known keyboard shortcuts (do not bind to these keys)
+const RESERVED_SHORTCUTS: Record<string, KeyCombo[]> = {
+  // macOS keybindings
+  mac: [
+    { key: 'q', meta: true }, // Quit
+    { key: 'w', meta: true }, // Close window
+    { key: 'h', meta: true }, // Hide
+    { key: 'm', meta: true }, // Minimize
+    { key: 'Tab', meta: true }, // App switcher
+    { key: ' ', meta: true }, // Spotlight
+    { key: 'F3', ctrl: true }, // Mission Control
+  ],
+  // Windows keybindings
+  win: [
+    { key: 'F4', alt: true }, // Close window
+    { key: 'Delete', ctrl: true, alt: true },
+    { key: 'Meta' }, // Start menu
+    { key: 'l', meta: true }, // Lock
+    { key: 'd', meta: true }, // Desktop
+    { key: 'Tab', meta: true }, // Task switcher
+  ],
+  // Linux keybindings
+  linux: [{ key: 'F4', alt: true }, { key: 'Delete', ctrl: true, alt: true }, { key: 'Meta' }],
+  // Browser-specific keybindings (cross-platform)
+  browser: [
+    { key: 't', ctrl: true },
+    { key: 'w', ctrl: true },
+    { key: 'n', ctrl: true },
+    { key: 'n', ctrl: true, shift: true },
+    { key: 't', ctrl: true, shift: true },
+    { key: 'p', ctrl: true, shift: true },
+  ],
+};
+const ALL_RESERVED_SHORTCUTS = [
+  ...RESERVED_SHORTCUTS.mac,
+  ...RESERVED_SHORTCUTS.win,
+  ...RESERVED_SHORTCUTS.linux,
+  ...RESERVED_SHORTCUTS.browser,
+];
+
+const attachmentFactory = (listenerFactory: KeyDownListenerFactory) => (element: KeyTargets) => {
```

Comment:
Talked to Jason and we both don't like the `...Factory` naming. In this case maybe just `getAttachment`? In some other cases like `keyDownListenerFactory` I think you can drop the `factory` just entirely from the name